solutions/gold/ioi-05-garden.mdx

```diff
@@ -0,0 +1,96 @@
+---
+id: ioi-05-garden
+source: IOI 2005
+title: 2005 - Garden
+author: Justin Ji
+---
+
+## Explanation
+
+Note that between any two disjoint rectangles, we can draw either a vertical
+or horizontal line between these two rectangles. We can consider every possible
+vertical and horizontal line, and find the two best rectangles between the
+lines.
+
+Let's first consider every vertical line. Consider fixing the two rows of our
+rectangle. If we can calculate the best rectangle perimeter for
+every prefix and suffix of our available columns, we can find our answer by
+sweeping over the splitting point. The implementation below iterates on the
+right column and then uses two pointers to find the best left column.
+Similar logic can be applied to find the best result for every horizontal line.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(l^{2}w + wl^{2})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	int l, w;
+	cin >> l >> w;
+	int n, k;
+	cin >> n >> k;
+	vector<vector<int>> grid(l + 1, vector<int>(w + 1));
+	for (int i = 0; i < n; i++) {
+		int x, y;
+		cin >> x >> y;
+		++grid[x][y];
+	}
+	const int INF = 1e9;
+	int res = INF;
+	for (int t = 0; t < 2; t++) {
+		vector<int> pref_min(w + 1, INF);
+		vector<int> suff_min(w + 1, INF);
+		for (int row_1 = 1; row_1 <= l; row_1++) {
+			vector<int> col_sum(w + 1);
+			for (int row_2 = row_1; row_2 <= l; row_2++) {
+				int ptr = 0, rect_sum = 0;
+				for (int col = 1; col <= w; col++) {
+					col_sum[col] += grid[row_2][col];
+					rect_sum += col_sum[col];
+					while (rect_sum >= k) {
+						rect_sum -= col_sum[ptr++];
+						if (rect_sum < k) {
+							rect_sum += col_sum[--ptr];
+							break;
+						}
+					}
+					if (rect_sum == k) {
+						int perimeter = 2 * (row_2 - row_1 + 1 + col - ptr + 1);
+						pref_min[col] = min(pref_min[col], perimeter);
+						suff_min[ptr] = min(suff_min[ptr], perimeter);
+					}
+				}
```

Comment:
all this w/o a single comment

Reply:
oops, will add some in a sec

Reply:
added a few comments, lmk if u think some more are warranted